src/net_processing.cpp

```diff
@@ -3988,6 +3988,51 @@ void PeerLogicValidation::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerLogicValidation::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
+    // If we have any extra block-relay peers, disconnect the youngest unless
+    // it's given us a block -- in which case, disconnect the second youngest
+    // instead (so we rotate our youngest block-relay peer).
+    int extra_block_relay_peers = m_connman.GetExtraBlockRelayCount();
+    if (extra_block_relay_peers > 0) {
+        // Pick the block-relay peer that we just connected, unless it gave us
+        // a block -- in which case disconnect the next-youngest block-relay
+        // peer.
+        std::pair<NodeId, int64_t> youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+        std::pair<NodeId, int64_t> second_youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+
+        m_connman.ForEachNode([&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            if (!pnode->IsBlockOnlyConn() || pnode->fDisconnect) return;
+            if (pnode->GetId() > youngest_peer.first) {
+                second_youngest_peer = youngest_peer;
+                youngest_peer.first = pnode->GetId();
+                youngest_peer.second = pnode->nLastBlockTime;
+            }
+        });
+        NodeId to_disconnect = youngest_peer.first;
+        if (youngest_peer.second > 0) {
+            // Our newest block-relay peer gave us a block; disconnect our
+            // second youngest.
+            to_disconnect = second_youngest_peer.first;
+        }
+        m_connman.ForNode(to_disconnect, [&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            // Make sure we're not getting a block right now, and that
```

Comment:
nit: I had to double-check that `nBlocksInFlight` can't be tricked with fake blocks. If you gonna update it, maybe just say "block with a valid header"?

Reply:
My worry was that `nBlocksInFlight` might contain just a hash for an announced block (pre-compact blocks) which we have no idea about. Then announcing any random hash would trick us.

Your current comment is sufficient to clarify that this is not possible.